utils/sharedcache/sharedcache_immutable.go

```diff
@@ -0,0 +1,232 @@
+package sharedcache
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/ARM-software/golang-utils/utils/commonerrors"
+	"github.com/ARM-software/golang-utils/utils/filesystem"
+	"github.com/ARM-software/golang-utils/utils/idgen"
+	"github.com/ARM-software/golang-utils/utils/parallelisation"
+)
+
+const (
+	partFileDescriptor     = ".part"
+	defaultCachedPackageID = "cb93fdbe-6c2e-4f7d-96ac-c422fc52618e "
+)
+
+type SharedImmutableCacheRepository struct {
+	AbstractSharedCacheRepository
+}
+
+type FileWithModTime struct {
+	filename string
+	modTime  time.Time
+}
+
+func NewSharedImmutableCacheRepository(cfg *Configuration, fs filesystem.FS) (repository *SharedImmutableCacheRepository, err error) {
+	abstractCache, err := NewAbstractSharedCacheRepository(cfg, fs)
+	if err != nil {
+		return
+	}
+	repository = &SharedImmutableCacheRepository{
+		AbstractSharedCacheRepository: *abstractCache,
+	}
+	return
+}
+
+func listCompleteFilesByModTime(ctx context.Context, fs filesystem.FS, entryDir string) (sorted []string, err error) {
+	err = parallelisation.DetermineContextError(ctx)
+	if err != nil {
+		return
+	}
+
+	// sort the files by mod time
```

Comment:
move this above the function declaration, it makes it seem like you are saying that Ls will sort by mod time